sharktank/sharktank/models/llm/testing.py

```diff
@@ -1,93 +1,187 @@
-        int(torch.max(seq_lens)), model.config.block_seq_stride
+    reference_model = TorchInstance(
+        reference_theta,
+        reference_config,
+        device=device,
+        prefill_bs=batch_size,
+        decode_bs=batch_size,
     )
-    token_ids = torch.randint(
-        low=0,
-        high=model.config.hp.vocab_size,
-        size=[batch_size, batch_seq_len],
-        dtype=torch.int32,
-        device=model.device,
+
+    pp_config = deepcopy(reference_config)
+    pp_config.parallelism_config = ParallelismConfig.default_config(
+        block_count=reference_config.hp.block_count,
+        pp=pipeline_parallelism_size,
     )
+    pp_theta = Theta(reference_theta.flatten())
+    pipeline_parallelize_llm_theta(pp_theta, pp_config.parallelism_config)
 
-    seq_block_ids = [
-        torch.arange(
-            batch_size * batch_seq_len // model.config.block_seq_stride,
-            device=model.device,
-        ).view(batch_size, -1)
-    ]
-    cache_state = model.cache.allocate(page_count=seq_block_ids[0].numel() + batch_size)
-    cache_state = [torch.rand_like(cache_state[0])]
-    return OrderedDict(
-        [
-            ("tokens", token_ids),
-            ("seq_lens", seq_lens),
-            ("seq_block_ids", seq_block_ids),
-            ("cache_state", cache_state),
-        ]
+    pp_model = TorchInstance(
+        pp_theta, pp_config, device=device, prefill_bs=batch_size, decode_bs=batch_size
     )
+
+    def make_llm_instance(model: TorchInstance):
+        page_sizes = llama_config_page_sizes(model.config)
+        page_count = minimum_required_kv_cache_page_count_for_batch(
+            tokens=tokens, config=model.config
+        )
+
+        return LlmInstance(
+            model_instance=model,
+            page_sizes=page_sizes,
+            block_seq_stride=model.config.block_seq_stride,
+            block_count=page_count,
+            kv_cache_dtype=dtype_to_serialized_name(model.config.kv_cache_dtype),
+            decode_topk_logits=None,
+        )
+
+    tester = LlmPerplexityCompare(
+        make_target_model=functools.partial(make_llm_instance, model=pp_model),
+        make_reference_model=functools.partial(
+            make_llm_instance, model=reference_model
+        ),
+        atol=0,
+        rtol=0,
+    )
+    tester.run_and_assert_close(tokens)
+
+
+def clip_llm_block_count(
+    theta: Theta, config: LlamaModelConfig, block_count: int
+) -> tuple[Theta, LlamaModelConfig]:
+    """Remove all trailing layers/blocks from the theta to align the desired block/layer count."""
+    assert (
+        config.pipeline_parallelism_size == 1 and config.tensor_parallelism_size == 1
+    ), "Not supported"
+
+    config = deepcopy(config)
+    config.hp.block_count = block_count
+    # Make sure block_count derivative values are recomputed.
+    config_as_props = config.to_properties()
+    del config_as_props["parallelism_config"]
+    del config_as_props["tensor_parallelism_size"]
```

Comment:
Why is this deletion needed?

Reply:
I guess specifically
```
del config_as_props["tensor_parallelism_size"]
```

is defensive programming. I will remove this one.